report/renderer/report-utils.js

```diff
@@ -187,6 +190,77 @@ class ReportUtils {
     return clone;
   }
 
+  /**
+   * Given an audit's details, identify and return a URL locator function that
+   * can be called later with an `item` to extract the URL of it.
+   * @param {LH.FormattedIcu<LH.Audit.Details.TableColumnHeading[]>} headings
+   * @return {{(item: LH.FormattedIcu<LH.Audit.Details.TableItem>): string|undefined}=}
+   */
+  static getUrlLocatorFn(headings) {
+    // The most common type, valueType=url.
+    const urlKey = headings.find(heading => heading.valueType === 'url')?.key;
+    if (urlKey) {
+      // Return a function that extracts item.url.
+      return (item) => item[urlKey]?.toString();
+    }
+
+    // The second common type, valueType=source-location.
+    const srcLocationKey = headings.find(heading => heading.valueType === 'source-location')?.key;
+    if (srcLocationKey) {
+      // Return a function that extracts item.source.url.
+      return (item) => {
+        const sourceLocation = item[srcLocationKey];
+        if (typeof sourceLocation === 'object' && sourceLocation.type === 'source-location') {
+          return sourceLocation.url;
+        }
+      };
+    }
+
+    // More specific tests go here, as we need to identify URLs in more audits.
+  }
+
+  /**
+   * Mark TableItems/OpportunityItems with entity names.
+   * @param {LH.Result.Entities|undefined} entityClassification
+   * @param {import('../../types/lhr/audit-result').Result} audit
+   */
+  static classifyEntities(entityClassification, audit) {
+    if (!entityClassification) return;
+    if (!(audit.details?.type === 'opportunity' || audit.details?.type === 'table')) {
+      return;
+    }
+
+    // If details.items are already marked with entity attribute during an audit, nothing to do here.
+    const {items, headings} = audit.details;
+    if (!items.length || items.some(item => item.entity)) return;
+
+    /**
+     * There is a common method to extract the url from all the items of a specific audit. That logic is encapsulated
+     * by getUrlLocatorFn.
+     */
```

Comment:
IMO this comment isn't necessary. It's reiterating the jsdoc comment for `getUrlLocatorFn`

Reply:
It's not just reiterating, I think this comment is the difference between me understanding what is going on and not 🤷

Reply:
I think in an earlier revision I had a (overly) verbose description as the comment (see below). I'd err on the side of having a comment if at least some people are finding the indirection confusing. 

```
/**
 * To avoid the downside of setting up a block of conditional statements here 
 * that determine if/where a URL is located within an item based on the audit
 * (and then switching between them later as we loop through items), we try to
 * create a URL locator function that can be called on each item to extract the URL.
 */
```

Reply:
Updated the comment.